### DISCUSSION TITLE
Newbie here

### DISCUSSION THREAD

**Discussion by GMPerfectZ** [TwitchIO/TwitchIO#370] [Questions] [answered]
I have 2 questions

1º Can I add a timer and iterations (like in routines), to avoid spamming a command?

2º I'm trying to count how many messages I've sent, but I'm having some issues.

Thanks in advance.

 `from twitchio.ext import commands
 from random import randint
 from time import sleep
 import random

 msg_counter = 0

 class Bot(commands.Bot):

    def __init__(self):
        # Initialise our Bot with our access token, prefix and a list of channels to join on boot...
        tag = ['!', '?']
        channels = ['Test_Channels']
        super().__init__(token='xxxxxxxx', prefix = list(tag), initial_channels= list(channels))

    async def event_ready(self):
        # We are logged in and ready to chat and use commands...
        print(f'Logged in as | {self.nick}')
        print(f'User id is | {self.user_id}')

    @commands.command()
    async def test(self, ctx: commands.Context):
        sleep(randint(3,6))
        test = ["test", f"test {ctx.author.name}"]
        await ctx.reply(random.choice(test)), await ctx.send(random.choice(test))
        global msg_counter += 1  #+= 1 with error
        print(msg_counter)
 bot = Bot()
 bot.run() `

**Comment by chillymosh** [comment, marked as answer]
Hi,

Please do not use time.sleep() as this is blocking the event loop. You should be using asyncio.sleep() if you need to pause a coroutine.
Globals are also a bad practice to use, especially in an OOP style framework. Simply create an attribute of the bot subclass.

I'm also not quite sure why you are calling `list()` on your prefixes when they are already a list `tag = ['!', '?']`
The same goes for channels.....

1. This is exactly one of the purposes of routines, as you can see in the examples in the documentation, you can add a time for which you want the routine to routinely run and how many (or infinite) iterations. https://twitchio.dev/en/latest/exts/routines.html
2. Create `self.msg_counter = 0` as an attribute in the Bot's `__init__` then in the command simply access and change that.
`self.msg_counter +=1`

Don't forget you can join the Discord support server for further help where examples will most likely have already been posted for this :)

**Reply by GMPerfectZ** [reply]
> Please do not use time.sleep() as this is blocking the event loop. You should be using asyncio.sleep() if you need to pause a coroutine.
Globals are also a bad practice to use, especially in an OOP style framework. Simply create an attribute of the bot subclass.

Thanks, I changed.

> I'm also not quite sure why you are calling list() on your prefixes when they are already a list tag = ['!', '?']
The same goes for channels.....

Actually, I'm using words I changed to "!" and "?" just to keep simple the code.

1º I couldn't make the routine code work the way I was planning so that's my I'm using commands. But now I'm looking to make it ignore to avoid spamming.

2º Thank you so much, it worked. <3

**Reply by chillymosh** [reply]
It shouldn't matter if ! and ? are words, they would still already be in a list when you define them as `tag = ["word1", "world2"]` which means you don't need to call list(). Same for `channels = ["chan1", "chan2"] this is already a list.

What are you trying to do with routines that the example does not show or explain? It should be fairly straightforward with the examples shown. I'm more than happy to give you a working example either here or on Discord.

**Comment by chillymosh** [comment]
For reference for others: the goal was to add a cooldown for the commands, which can be achieved using the Cooldown decorator.
https://twitchio.dev/en/latest/exts/commands.html#twitchio.ext.commands.Cooldown
